pkg/aws/metadata/handler_proxy.go

```diff
@@ -0,0 +1,57 @@
+// Copyright 2017 uSwitch
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package metadata
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"regexp"
+
+	"github.com/gorilla/mux"
+)
+
+type proxyHandler struct {
+	metadataURL   *url.URL
+	reverseProxy  http.Handler
+	allowedRoutes *regexp.Regexp
+}
+
+func (p *proxyHandler) Install(router *mux.Router) {
+	p.reverseProxy = httputil.NewSingleHostReverseProxy(p.metadataURL)
+	router.Handle("/{path}", adapt(withMeter("proxy", p)))
+}
+
+type teeWriter struct {
+	http.ResponseWriter
+	status int
+}
+
+func (w *teeWriter) WriteHeader(statusCode int) {
+	w.status = statusCode
+	w.ResponseWriter.WriteHeader(statusCode)
+}
+
+func (p *proxyHandler) Handle(ctx context.Context, w http.ResponseWriter, r *http.Request) (int, error) {
+	route := mux.Vars(r)["path"]
+	if p.allowedRoutes.MatchString(route) {
+		writer := &teeWriter{w, http.StatusOK}
+		p.reverseProxy.ServeHTTP(writer, r)
+		return writer.status, nil
+	} else {
+		return http.StatusForbidden, fmt.Errorf("request blocked by allowedRoutes pattern %q: %s", p.allowedRoutes, route)
```

Comment:
Is `403` the correct HTTP status code to return?
From [RFC 7231](https://tools.ietf.org/html/rfc7231#section-6.5.3):
>  An origin server that wishes to "hide" the current existence of a
> forbidden target resource MAY instead respond with a status code of
> 404 (Not Found).

Reply:
the request is not allowed by (does not match) the `allowed-routes` regular expression, so the request is forbidden by the configuration. which status code should be returned in this case?

Reply:
I don't think there's a "correct" answer. Either would be valid.

Reply:
Given that RFC maybe we should favour 404?

We should still log that it was blocked via a warning because of the path regexp (so that its easy for operators to record/debug) but 404 feels like it leaks less information to clients.

@uswitch/cloud any other thoughts? Feels pretty close to a coin toss though 😄

Reply:
I have no issues with changing the status code here since I went with my gut and `MAY` means it's up to us. Flip that coin.